platforms/chibios/chcore_overrides.c

```diff
@@ -0,0 +1,17 @@
+// Copyright 2023 Nick Brassel (@tzarc)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+// Weak implementation of NMI_Handler_Original, so that if ChibiOS doesn't implement it, it becomes a no-op and safe
+// for linking.
+__attribute__((weak)) void NMI_Handler_Original(void) {}
+
+// Re-route to the original NMI_Handler, whilst allowing QMK to perform specialisations.
+void NMI_Handler(void) {
+    NMI_Handler_Original();
+
+#ifdef WEAR_LEVELING_DRIVER_EMBEDDED_FLASH
+    // Handle wear-leveling errors
+    extern void NMI_Handler_EFL(void);
+    NMI_Handler_EFL();
+#endif // WEAR_LEVELING_DRIVER_EMBEDDED_FLASH
+}
```

Comment:
Apparently ChibiOS can itself use the NMI handler in the `ARMv6-M` (Cortex-M0, except RP2040) and `ARMv6-M-RP2` (RP2040) ports, depending on the value of the `CORTEX_ALTERNATE_SWITCH` option (when `CORTEX_ALTERNATE_SWITCH` is `FALSE`, the NMI handler is used for preemption; when `CORTEX_ALTERNATE_SWITCH` is `TRUE`, the PendSV handler is used instead).  The default value for that option is `FALSE` for both of those ports (so NMI is used).  And that code is used in the exit path of all interrupt handlers.

According to the [ChibiOS ARMv6-M port documentation](http://www.chibios.org/dokuwiki/doku.php?id=chibios:documentation:books:rt:port_armv6m), using PendSV for context switch has a downside of reserving the priority level 0 for that purpose (priority level 0 cannot be used for any other interrupts); there is also one more instruction (`cpsie i`) in the interrupt exit path (which is probably not noticeable except in microbenchmarks).  Adding extra code to the NMI handler when it is actually used for context switch would be much worse though.

So ideally this NMI handler replacement should be done only when actually needed.  Just disabling the code for Cortex-M0 is not an option because of STM32G0B1xx, STM32G0C1xx (these chips are M0+, but have ECC flash and USB).  Requiring `#define CORTEX_ALTERNATE_SWITCH TRUE` for that configuration may be reasonable though.

Reply:
Actually it seems that the NMI handler from ChibiOS cannot coexist with any other kind of NMI usage, because it unconditionally assumes that any NMI was caused by the ChibiOS `__port_exit_from_isr` code, therefore it would do some completely wrong things if the NMI was caused by anything else.  Therefore the only correct option for the ARMv6 ports would be to require `#define CORTEX_ALTERNATE_SWITCH TRUE` if the custom NMI handler is required, and the `NMI_Handler_Original` redefinition can just be dropped.

BTW, some other Cortex-M0\[+\] chips that may need to be checked for breakage here are KL2x (MKL26Z64, Teensy LC), NUC123 (Ducky boards), HT32F523xx (Anne Pro 2); SN32 and LPC11Uxx are M0 too.  And GD32VF103 is not even ARM, so it does not have the NMI handler like that.

Reply:
Have limited the code to L4xx for now -- as it's an ARMv7 part the issues with NMI shouldn't occur as the ChibiOS code doesn't use NMI for context switching.